model/error/event.go

```diff
@@ -91,6 +107,10 @@ type Event struct {
 
 	Experimental interface{}
 	data         common.MapStr
+
+	Metadata metadata.Metadata
+
+	decoder decoder
```

Comment:
I think this should be `*decoder`, and `func (d decoder) Decode(...)` should instead be a method on the pointer type. Otherwise there'll be a lot of unnecessary copying.

Reply:
I had forgotten that this Event type is also constructed directly by processor/otel, so perhaps referencing decoder itself here is not the right thing to do after all. It's not really a problem at the moment, as RUM data doesn't go through processor/otel, but it's still a bit ugly.

Another couple of options:
 1. Have Decoder wrap Event with another type that (RUMEvent?) that has the sourcemap-related bits embedded.
 2. Introduce middleware for post-processing Transformables. The middleware would encapsulate the sourcemap/stacktrace configuration, and the Transformable interface would remain simple so we wouldn't have to keep plumbing the config through all the different layers.

Reply:
I had thought of `1`. 
Can you elaborate on the middleware idea? With post-processing you mean to work on the `beat.Event` types?

Reply:
It could be done like that (it would basically be a pipeline processor?), or we could have something like this:

In `package publish`:
```go
package publish

type TransformerMiddleware func(model.Transformable)
```

(Publisher could be configured with a TransformerMiddleware, calling it for each Transformable before finally calling the Transform method.)

In the `model/error` package you'd have something like this:

```go
package error

func NewErrorStacktraceMiddleware(sourcemapStore *sourcemap.Store) func(model.Transformable) model.Transformable {
    return func(t model.Transformable) {
        switch t := t.(type) {
        case *Event:
            // Use sourcemapStore etc. to update the event
        }
    }
}
```

It seems kinda dirty to me how `Transform` modifies the Stacktrace objects. Perhaps it would be better to operate on the generated `beats.Event` output instead.

Reply:
Ok. I think that this solution will spread the number of places where data is juggled, and add another thing you have to "remember" when adding a new event type (ie another switch case). I feel it is not worth it...

I like the idea of injecting the config settings in the type (either as wrapper with the configs + the actual event, or the fields directly embedded in the event); but for the sourcemap I think the option less bad is to pass it as argument to `Transform` from the publisher.

If we later have the need for more caches (ie. api keys per service - can't picture how implementation will look like, but could need something like it), we then can embed them all in a big `state`/`context` object (the word "context" has being seized by go's package, but semantically is accurate: something that affects event transformation other than the events themselves). wdyt?

Reply:
I agree that the second middleware approach isn't great, and shunting this work off to a "pipeline processor" isn't great either, but I don't want to give up yet! ;)

Context is for carrying request/call-specific context. In this case, sourcemap config is not specific to the individual call context. It's configuration that applies to _all_ calls, hence why I don't like having to pass it around all the time.

So, one more idea: we stop making the model object types responsible for transforming themselves, and introduce a separate Transformer. This is the thing that would be configured, rather than attaching transformation config to the model object.

With this approach the model packages would only be concerned with decoding and providing a common object representation based on the intake API. Reducing to these responsibilities, we could more easily auto-generate the struct types and decoding from the/a spec, and we would just have to hand write the transformation to events.

(If you want to get back to your original change, I can also take a stab at this later.)

Reply:
ok... i ended implementing half of the idea (actually i implemented all of it, and then reverted the last part), i found it a bit weird that the decoding happens on each event (multiple packages) and the transformation all in one place.

I also got back to the original decoder functions - I don't see the point of having 1 interface + 5 additional types (1 per implementation) when it can be done with normal functions...

Reply:
For bystanders: I had a play with the branch, and came up with this: https://github.com/jalvz/apm-server/compare/lay-rum-ground...axw:lay-rum-ground

Reply:
The last couple of days I have been looking into alternatives regarding decoding and validations allowing us to clean up the currently used decoding and schema validation logic. While I did not touch the overall beater/API handling structure, this might still be interesting in the context of your discussion here. I make sure to send a wrap up until next week.

Reply:
> Reducing to these responsibilities, we could more easily auto-generate the struct types and decoding from the/a spec, and we would just have to hand write the transformation to events.

This is exactly what I've looked into.